### PR TITLE
Removed `max_length` from being mandatory within `generate`.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Moving on to fully using `StoppingCriteria` for `greedy` and `sample`
modes.
- `max_length` still used for `beam_search` and `group_beam_search`
(Follow up PR)
- Fixes a bug with MaxLengthStoppingCriteria (we should stop as soon a
we hit the max_length, the comparison needs to be or equal, that affects
the tests).
- Added options to use `logits_processor` and `stopping_criteria`
directly within `generate` function (so some users can define their own
`logits_processor` and `stopping_criteria`).
- Modified the backward compat tests to make sure we issue a warning.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@patrickvonplaten @LysandreJik

### CODE REVIEW

File: src/transformers/generation_stopping_criteria.py [hunk 1]
```diff
@@ -83,15 +83,16 @@ class StoppingCriteriaList(list):
...
-    found = False
-    for stopping_criterium in stopping_criteria:
-        if isinstance(stopping_criterium, MaxLengthCriteria):
-            found = True
-            if stopping_criterium.max_length != max_length:
-                warnings.warn(
-                    "You set different `max_length` for stopping criteria and `max_length` parameter", UserWarning
-                )
-    if not found:
+    smax_length = stopping_criteria.max_length
```
Comment (patrickvonplaten) [inline src/transformers/generation_stopping_criteria.py:94]: why `smax_length`?

File: src/transformers/generation_stopping_criteria.py [hunk 2]
```diff
@@ -83,15 +83,16 @@ class StoppingCriteriaList(list):
...
+    if smax_length is not None and smax_length != max_length:
+        warnings.warn("You set different `max_length` for stopping criteria and `max_length` parameter", UserWarning)
+    elif smax_length is None:
         stopping_criteria.append(MaxLengthCriteria(max_length=max_length))
```
Comment (patrickvonplaten) [inline src/transformers/generation_stopping_criteria.py:98]: I think it's more in line with the libraries' philosophy if we could actually return the `stopping_criteria` here so that this line:

```python
validate_stopping_criteria(stopping_criteria, max_length)
```

becomes this line:
```python
stopping_criteria = validate_stopping_criteria(stopping_criteria, max_length)
```

Since the function `validate_stopping_criteria(...)` changes one of its argument, I think it's more understandable if the function returns the changed variable - otherwise one might assume that the function only checks something but doesn't actually change anything - what do you think?

File: src/transformers/generation_stopping_criteria.py [hunk 3]
```diff
@@ -83,15 +83,16 @@ class StoppingCriteriaList(list):
     def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
         return any(criteria(input_ids, scores) for criteria in self)
 
+    @property
+    def max_length(self) -> Optional[int]:
```
Comment (patrickvonplaten) [inline src/transformers/generation_stopping_criteria.py:87]: I like it!

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -484,13 +485,11 @@ def _expand_inputs_for_generation(
         return input_ids, model_kwargs
 
     @staticmethod
-    def _init_sequence_length_for_generation(
-        input_ids: torch.LongTensor, max_length: int
-    ) -> Tuple[torch.Tensor, torch.Tensor, int]:
+    def _init_sequence_length_for_generation(input_ids: torch.LongTensor) -> Tuple[torch.Tensor, torch.Tensor, int]:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:488]: I think [this PR](https://github.com/huggingface/transformers/pull/11331) will resolve the issue with `max_length` so that we can maybe merge it first - sorry for the merge conflict :-/

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -689,6 +691,8 @@ def generate(
         num_beam_groups: Optional[int] = None,
         diversity_penalty: Optional[float] = None,
         prefix_allowed_tokens_fn: Optional[Callable[[int, torch.Tensor], List[int]]] = None,
+        logits_processor: Optional[LogitsProcessorList] = None,
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:694]: Hmm, not super happy about this - is it really needed to define `logits_processor` outside of `generate()` ? What would be the use case for this? We would then also have to add `logits_warper` to it. Also what do we do then if a user say passes `no_repeat_ngram_size` and passes a `logits_processor` with `no_repeat_ngram_size`? Which one should we take? 

=> I would prefer to not add them to the arguments of `generate()`. What do you think?

File: src/transformers/generation_stopping_criteria.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/generation_stopping_criteria.py]: We can, but I feel like we should change the function to make it functional (recreate a new obejct) if I go that way.
It's super dangerous to have a function that looks functional but is not.

I switched to a functional function that returns a new stopping_criteria_list, it should be mostly very low cost, and much less surprising in behavior

File: src/transformers/generation_utils.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/generation_utils.py:671]: I've had a discussion where users where actually using custom LogitsProcessor (like penalize certain tokens and whatnot, a bit like bad_words, but seemed even more specialized). 

It made sense, but this could very well go in another PR (if we even desire it)

Comment (Narsil) [reply src/transformers/generation_utils.py:671]: And the behavior is all logits processor will run, it's the caller's responsability to know what is in it IMO. Like stopping_criteria.

Comment (Narsil) [reply src/transformers/generation_utils.py]: I removed it from this PR.

**Review by patrickvonplaten — APPROVED** [review]
Great! 

Could we run the Bart and T5 slow tests, just to make sure we didn't increase or decrease the allowed max_length by 1:

```
RUN_SLOW=1 pytest tests/test_modeling_bart.py
RUN_SLOW=1 pytest tests/test_modeling_t5.py
```

Those test have some aggressive beam search tests that are very sensible to the `max_length` param

File: src/transformers/generation_stopping_criteria.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/generation_stopping_criteria.py]: Renamed to stopping_max_length (just avoiding shadowing the argument name)

**Comment by Narsil** [comment]
All tests pass, the max_length was actually a bug hidden by  `while cur_len < max_length` that was still in there.
The bart tests (at least) caught it automatically and enabled me to change it to the correct comparison !

**Comment by patrickvonplaten** [comment]
> All tests pass, the max_length was actually a bug hidden by `while cur_len < max_length` that was still in there.
> The bart tests (at least) caught it automatically and enabled me to change it to the correct comparison !

Great! Feel free to merge then!
